### ISSUE TITLE
New contributor guide

### ISSUE THREAD

**Issue by dkypuros** [kubernetes/website#20485]
**This is a Feature Request**

**What would you like to be added**
See summary at bottom of my page. I would like a better understanding of "what Git and Github commands do I need to know to do(start) my first PR (basic)?"

**Why is this needed**
Given I start [here](https://kubernetes.io/docs/contribute/), and hile following the third-party links to, in this case "[Git](https://git-scm.com/)", I don't have a reference point of what I need to get out of the tutorial process. No pun intended.

**Comments**
Here all my notes after following the official Kubernetes > Documentation > Contribute link off the main website. This section of my notes just covers the first link "Git" under "Getting Started". That was ~2.5  hours of watching and learning.

#### Part 1 - Simple
Setting up the initial git control structures. This creates a folder for the project files, and the "control files" that capture the historical elements. History snapshots of images, documents, etc.
```
$ git init myproject
```

This is the command that "notices" the files, and puts them into a holding zone, ready to be permanently committed.
```
$ git add .
```

This is the command that "notices" the files, and puts them into a holding zone, ready to be permanently committed. A commit, permanently records a snapshot of the files as they exist at a current point in time. Is this like a "time-stamp" for whatever is #1 added via the previous command (is that the context for the commit) #2 This commit only happened on my computer.
```
$ git commit -m"Importing all the code"
```

#### Part 2 - Collaborative Git / Distributed Git
"Bring back their work - to the home base (rebase?)" to have it incorporated into the primary copy of the project (main branch?).

People having their own local copy, adding changes, but needing to unify (...or Merge) their local copy with the central copy of the project.

Working in parallel and at different times - Doug and Dana.
Doug can work on his "master branch" - an arm of the project. (Question: is that Doug's "myproject" running on his local machine?)

//Doug's machine
```
$ git checkout master
$ git commit -a -m"My new logo"
$ git push //This has the network component, sending changes to a centralized server
```

//Dana's machine
```
$ git checkout -b danasfeature  //why isn' she doing "git checkout master"?
$ git commit -a -m"My feature code"
$ git push origin danasfeature //Now Dana is bringing her changes into the central server copy. What is "origin"? Why didn't Doug have to that?
```

//Doug's machine
```
$ git pull   //Retrieve copy of centralized versioned files onto Doug's machine.
$ git merge danasfeature //In this context Doug is pulling together both of their work "merging" it all together. Brings all the code into a unified set.
```

#### Part 3 - Advanced Git
Glue together small programs.
Piece by piece adding of small documents
```
$ git add -p MyReport.markdown  //add some (specific) of the pieces selected.
$ git commit -m"Added latest statistics"
$ git log --graph --decorate --abbrev-commit --all --pretty=online //historical graphical representation. Show when people branched files, and merged them back together.
```

#### Part 4 - How to install, setup, configure, and use git from CLI for the first time.
Where you get git. Linux, windows, and mac. https://git-scm.com/downloads. Need username and email.

These are not credentials, just for record keeping for things like commits. "usern.name" is the key we're setting, and "Dana Devops" is the value.

Do this before your first commit. Commit uses this information along with the commit message.
```
$ git config --global username.name "Dana Devops"  
$ git config --global username.email "danadevops@gmail.com"  
```

Create your first git repository.
```
$ git init project1
$ cd project1

$ create file1.txt

$ git add file1.txt  //Add signals, we want the file to participate in the next commit. "add" allows use of "commit"

$ git commit -m"My first commit"   //finish by making the changes a permanent record; recording a given state of the file at a point in time. "-m" means, commit "message"
```

Fedora 31 experience
These steps: Records a historical event in the "Git version control system"
```
[davidkypuros@localhost ~]$ git init project1
...terminal output...
Initialized empty Git repository in /home/davidkypuros/project1/.git/


git commit -m"My first commit"
...terminal output...
[master (root-commit) 2317ac4] My first commit
 Committer: David Kypuros <davidkypuros@localhost.localdomain>
Your name and email address were configured automatically based
on your username and hostname. Please check that they are accurate.
You can suppress this message by setting them explicitly. Run the
following command and follow the instructions in your editor to edit
your configuration file:

    git config --global --edit

After doing this, you may fix the identity used for this commit with:

    git commit --amend --reset-author

 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 file1.txt



$ git config --global username.name "David Kypuros"  
$ git config --global username.email "davidkypuros@localhost.localdomain"
```

#### Part 5 - Quick Wins with Git
* Precise enough to follow the "content" inside a file, not just a file
* Staging provides a way to "elect" which files "Opt in" to the next commit. This prevents committing files that you know you didn't change, to avoid accidents.
* Git doesn't need people to "lock" a file to make edits. It's open. It smart enough to handle multiple changes to content from different people/sources via merges.
* Exchange of code actually become conversations. When you bring in a bit of code into a code base, it's an opportunity to talk and share. It's not an outright rejection, but rather a request for small changes and tuning to make them acceptable.
* Focus on the people not the tools. "Make Git the smallest, lightest, most background piece you possibly can, and focus on what the objective is."" Writing good documents, software or GUI". Require minimal interaction.
* Journal, not backup. version control History rather than 1 big blog of changes at the end of the day. This "inter-day" set of activities is more helpful. It's like "Tevo" recording everything important you did to a file etc.
* Git works entirely offline. Commits, branches, merges, re-base's, resetting history, searching historical changes. Push and Pull are a minimal set of Git's vocabulary.

=========================================================

### Summary / Findings / Things Learned
1. I didn't know the squares were in a sequence. (1-4) [Here is what I mean - here](https://git-scm.com/doc).
2. Off the [main page](https://git-scm.com/) I didn't know where to start. I guessed "easy to learn"
3. I still don't know if I have all the information needed to move on to "learning GitHub" only using the link from [Kubernetes Docs](https://kubernetes.io/docs/contribute/).
4. I didn' realize the 4 videos would be as valuable as they were. I was surprised with a lot of rich content, when I stuck it out past the first video. The last video touching on the philosophy of git was very exciting. Favorite quote "Exchange of code actually become conversations".
5. Wondering where to go from here.
6. The big question, what Git and Github commands do I need to know to do(start) my first PR (basic).

###Wondering where to go from here?
* [Visual Cheat Sheet](https://ndpsoftware.com/git-cheatsheet.html#loc=workspace) More like a reference area for folks already very familiar with git etc.
* [GitHub Cheat Sheet](https://github.github.com/training-kit/downloads/github-git-cheat-sheet/) Is actually very user friendly, and adds nicely to the 4 training videos. The glossary section is vey helpful for a level set of names used throughout the videos.

###Items I haven't learned about yet from the videos etc.
1. No hands on learning of branch, checkout, merge.

###What might be nice to have
1. Visual flow diagram - Doug & Dana
2. Example diagram showing how the concepts / definitions work. Tree, Branch, Master, Merge...

## What are the commands I need to do a basic PR? That is the question.

| Category / Basic Skill | Where                   | Quick Name   | git | Commands I’ve Learned                                         | Still missing.                                                |
|------------------------|-------------------------|--------------|-----|---------------------------------------------------------------|---------------------------------------------------------------|
| Create repositories    | https://git-scm.com/doc | The 4 videos | git | init myproject                                                | Hands-on example included in the training video.              |
| Make changes           | https://git-scm.com/doc | The 4 videos | git | add .                                                         | Hands-on example included in the training video.              |
| Make changes           | https://git-scm.com/doc | The 4 videos | git | commit -m"Importing all the code"                             | Hands-on example included in the training video.              |
| Branches               | https://git-scm.com/doc | The 4 videos | git | checkout master                                               | Little explanation to take with me to a hands-on environment. |
| Synchronize changes    | https://git-scm.com/doc | The 4 videos | git | push                                                          | Little explanation to take with me to a hands-on environment. |
| Synchronize changes    | https://git-scm.com/doc | The 4 videos | git | push origin danasfeature                                      | No explanation of Origin.                                     |
| Synchronize changes    | https://git-scm.com/doc | The 4 videos | git | merge danasfeature                                            | Little explanation to take with me to a hands-on environment. |
| Make changes           | https://git-scm.com/doc | The 4 videos | git | log --graph --decorate --abbrev-commit --all --pretty=online  |                                                               |
| Configure tooling      | https://git-scm.com/doc | The 4 videos | git | config --global username.name "Dana Devops"                   |                                                               |
| Configure tooling      | https://git-scm.com/doc | The 4 videos | git | config --global username.email "danadevops@gmail.com"         |                                                               |
| Synchronize changes    | https://git-scm.com/doc | The 4 videos | git | pull                                                          |                                                               |

**Comment by celestehorgan** [comment]
So, when I initially re-did the Contribute section, my goal was to reduce duplication. One of the decisions I made to do this was to reduce the amount of Git-related content. Git is a third party tool, and knowing it is a prerequisite of contributing, similar to a developer needing to know Go before contributing to the codebase. I think I've taken the wrong stance on this, however.

I think we can do one of the following re: teaching Git:
1) Write and host a _really good and complete_ intro to Git tutorial
2) Offload writing/hosting a really good Git tutorial onto third party documentation and link to it.

From this issue + discussion from @kbarnard10 in the SIG-Docs meetings, it sounds like what we want is strategy #1.

I'm willing to take a stab at writing/organizing this if we're willing to support and maintain a full-fledged Git tutorial for the long run. :)

**Comment by dkypuros** [comment]
> So, when I initially re-did the Contribute section, my goal was to reduce duplication. One of the decisions I made to do this was to reduce the amount of Git-related content. Git is a third party tool, and knowing it is a prerequisite of contributing, similar to a developer needing to know Go before contributing to the codebase. I think I've taken the wrong stance on this, however.
> 
> I think we can do one of the following re: teaching Git:
> 
> 1. Write and host a _really good and complete_ intro to Git tutorial
> 2. Offload writing/hosting a really good Git tutorial onto third party documentation and link to it.
> 
> From this issue + discussion from @kbarnard10 in the SIG-Docs meetings, it sounds like what we want is strategy #1.
> 
> I'm willing to take a stab at writing/organizing this if we're willing to support and maintain a full-fledged Git tutorial for the long run. :)

I think we will need to host some kind of git tutorial. I found myself guessing where to go, once I left the kubernetes docs contribute site. 

 I'm thinking we start with the smallest amount of git skills one needs to do a basic PR on SIG-DOCS. We can add a "full fledged" tutorial as a part two. Part 1 might be targeted for about 30-40 min. Part 2 might targeted for about 2 hours. I mention time, only because the longer I spent learning git, without being able to submit a PR, the more demotivated I became to continue.

PS
I'm almost finished recording all my notes for learning GitHub via links provided on [docs](https://lab.github.com/).

**Comment by celestehorgan** [comment]
@dkypuros So, thinking more about this, I want to ask one question: what would creating and hosting our own Git tutorial give us that a well-curated set of links wouldn't?

By my estimates, this would be a tutorial about 5-10 *dense* pages in length at least. There's already lots of other tutorials online that with careful curation could serve the community just as well without us having to create and maintain content. 

What's the value add for us doing it ourselves?

**Comment by dkypuros** [comment]
What's the value add for us doing it ourselves?
==========================================

Value: More control over key content needed to do a PR for docs
----------------------------------------------------------------------------------------
@celestehorgan Great question. I don't know the answer, but I kinda feel like once we tell a kubernetes-docs visitor to click here and go learn somewhere else, we don't really know what happens when they leave [the site](https://kubernetes.io/docs/contribute/) (e.g. to go learn git/github), and [return to the site to ](https://kubernetes.io/docs/contribute/#your-first-contribution). 

Did the visitor learn what they needed to learn in order to continue to the [next step of contributing](https://kubernetes.io/docs/contribute/#your-first-contribution)? 

Value: Better examples geared toward Kubernetes
-----------------------------------------------------------------------
The example tutorial from GitHub is focused on a "slideshow project" rather than finding a first issue in Kubernetes docs. The learning happens around a different project / task.

I haven't posted my notes from the "GitHub" learning process yet, but I will share GitHub's video training explaining "how to do a PR", doesn't exactly mirror the Kubernetes PR flow. GitHub videos assume a very very easy model of PRs. 

Value: Cohesive tutorial across git->github->kubernetes/docs-website
-------------------------------------------------------------------------
It might be nice to show our site visitors: (1.) Here are the basics of Git. (2) Here are the basics of GitHub. (3). Here is the basics of a Kubernetes-DOCS PR flow. 

A curated list, is also good. It's less work. Once we host our own git / and github tutorial we're tied to updating that tutorial for ever. So I see both sides.

**Comment by dkypuros** [comment]
@celestehorgan 

I just did a check, and found on linux, it takes 22 total commands to complete a PR (plus some work in a few GUIs). Of the 22 commands, there are actually only 11 unique commands to learn.

Most people will only need know 11 commands to complete a PR.
======================================================
$ hugo
$ hugo server
$ git config
$ git clone
$ git remote
$ git fetch
$ git checkout
$ git status
$ git add
$ git commit
$ git push

## GUI's Prep work
====================================================

* [Create GitHub account](https://github.com/)
* Fork [kubernetes/website](https://github.com/kubernetes/website)
* [Sign CLA](https://github.com/kubernetes/community/blob/master/CLA.md#the-contributor-license-agreement)
* [Download Hugo extended](https://github.com/gohugoio/hugo/releases/download/v0.69.2/hugo_extended_0.69.2_Linux-64bit.tar.gz) - [and install](https://gohugo.io/getting-started/installing)


## Install Git and Hugo on workstation
========================================
$ sudo dnf install git-all
$ tar tvf hugo_extended_0.69.2_Linux-64bit.tar.gz]
$ tar -xvzf ~/Downloads/hugo_extended_0.69.2_Linux-64bit.tar.gz
$ ./hugo version
$ which hugo
$ hugo version
$ hugo
$ hugo server

## Git Origin / Upstream / Local
========================================
$ git config --global user.name "David Kypuros"
$ git config --global user.email "davidkypuros@gmail.com"
$ git clone https://github.com/dkypuros/website.git
$ git remote add upstream https://github.com/kubernetes/website.git
$ git remote -v
$ git fetch origin
$ git fetch upstream
$ git checkout -b new_architecture_branch upstream/master
$ git status
$ git add static/images/docs/components-of-kubernetes.png

## Test the changes locally
========================================
$ hugo
$ hugo server

## Git Commit
========================================
$ git commit -m "Updating Kubernetes architectural diagram image in Docs to reflect standard Kubernetes icons."
$ git push origin architecture

## GUI Submit Pull Request
========================================
* [kubernetes/website](https://github.com/kubernetes/website/) New Pull Request > compare across forks > head repository (select your fork) | Compare drop-down menu, select your branch = Create Pull Request button

**Comment by takeoffnyc** [comment]
Duplicate of #

**Comment by celestehorgan** [comment]
cc: @zacharysarah

**Comment by sftim** [comment]
@takeoffnyc can you clarify https://github.com/kubernetes/website/issues/20485#issuecomment-621502220? (which issue does this duplicate?)

**Comment by sftim** [comment]
How well does https://www.kubernetes.dev/docs/guide/#contributor-guide match the outcome that this issue was calling for?

**Comment by takeoffnyc** [comment]
Mentioned how
